onset_maproom/onset_indices.py

```diff
@@ -35,6 +36,62 @@ def onset_indices(
     onset_ids = np.unique(first_rainy[onset_ids] + onset_ids)
     return onset_ids
 
+def onset_index(
+    data, data_flag, wlength,      wthresh,       flagged_thresh, fwlength, false_search, coord_index
+):
+    """Finds the first window of size wlength where data exceeds wthresh,
+    with at least flagged_thresh count of flagged data (by data_flag),
+    not followed by a false_window of size fwlength of unflagged data, for the following false_search indices
+    returns the index of first flagged data in that window
+    (how much wood would a woodchuck chuck if a woodchuck could chuck wood?)
+    """
+    flagged = data > data_flag
+
+    window = (
+        data.rolling(**{coord_index: wlength}).sum() > wthresh
+    ) & (
+        flagged.rolling(**{coord_index: wlength}).sum() > flagged_thresh
+    )
+
+    unflagged = ~flagged
+    false_window = (
+        unflagged.rolling(**{coord_index: fwlength}).sum() == fwlength
+    )
+
+#Note that rolling assigns to the last position of the window
+    false_window_ahead = (
+        false_window.rolling(**{coord_index: false_search}).sum().shift(**{coord_index: false_search*-1}) != 0
+    )
+
+    flagged_within_onset = (
+        flagged
+    ) & (
+        (window & ~false_window_ahead).rolling(**{coord_index: wlength}).sum().shift(**{coord_index: 1-wlength}) != 0
```

Comment:
I'm having trouble understanding this line.  I understand that `window & ~false_window_ahead` is True on dates that are at the end of a wet spell and not followed soon by a dry spell. I understand that the shift moves the marker from the end of the wet spell to the start. But I don't understand how `rolling` is being used.

The first few values of the rolling sum are always `NaN` because there isn't a full window to look at, and `NaN != 0`, so this expression comes out to True on some dates that aren't anywhere near a wet spell.

Reply:
Assuming we are looking at a window of 3-day long wet spell, to simplify explanation, the first wet day can be the 3rd, 2nd or 1st day of any of the 3 spells it could be in. So the rolling.sum allows to do a "OR" on this given day being in one of those spells. Then shift moves the marker etc...

As for the 2nd comment, I did get onset dates way away from the beginning of times so I haven't checked but I suspect the 1-wlength shift removes the heading NaNs you are talking about.

Reply:
This could be a lot more simple if assuming input to be daily rainfall

Reply:
If assuming daily rainfall is going to allow you to simplify it, then I won't spend time trying to understand this.

Possibly related, although I'm not sure because I still don't understand this entirely... you define a wet spell as a period of length x containing at least y wet days and total rainfall of at least z, and then you find the first wet day of the first wet spell. I wonder if it would be simpler to add to the definition of a wet spell the requirement that it must start with a wet day.

Reply:
That would change things a bit as an earlier non-wet day could contribute to making a wet spell that would otherwise not be reaching the wet spell threshold. That's probably not many cases but possible. We are probably in that situation because that piece came as an afterthought. Once I wrote the algo to find the first wet spell, I was going to make that period of time to be the onset date, but then I was told to pick the 1st wet day within it.

But so... are we agreeing to move away from the general case and assume daily rainfall as input?

Reply:
I see. You're right, it wouldn't be equivalent.

Yes, let's assume daily rainfall.